src/socket/tcp.rs

```diff
@@ -936,6 +936,17 @@ impl<'a> TcpSocket<'a> {
             }
         }
 
+        // If a FIN arrives before the data, we should wait until all data arrived.
+        // (A slight improvement would be to remember that we got the FIN already.)
+        let window_start = self.remote_seq_no + self.rx_buffer.len();
+        if repr.control == TcpControl::Fin && repr.seq_number > window_start {
+            net_debug!("{}:{}:{}: FIN arrived before unseen data \
+                        ({} > {}), will send challenge ACK",
+                       self.meta.handle, self.local_endpoint, self.remote_endpoint,
+                       repr.seq_number, window_start);
+            return Ok(Some(self.ack_reply(ip_repr, &repr)))
+        }
```

Comment:
No, I don't think this is the right fix. It is discarding data arriving in the segment with FIN set, which is not necessary, and sending a challenge ACK, which is again not necessary because such an ACK will be sent once the currently missing segment will be received.